### PR TITLE
Love for remote debugging

### LINKED ISSUE #82: Write tests for remote debugging support

### LINKED ISSUE #141: Foreman / Passenger - address is already in use
I've followed the tutorial and added this to `config/development.rb`:

``` ruby
require 'byebug'
Byebug.wait_connection = true
Byebug.start_server('localhost', 4000)
```

When starting foreman, I get the error below. I've tried many different ports, the error message is always the same =(.

```
foreman start
11:05:01 web.1    | started with pid 1313
11:05:01 worker.1 | started with pid 1314
11:05:01 clock.1  | started with pid 1315
11:05:03 web.1    | =============== Phusion Passenger Standalone web server started ===============
11:05:03 web.1    | PID file: /Users/michaelimstepf/Dev/web-app/tmp/pids/passenger.5000.pid
11:05:03 web.1    | Log file: /Users/michaelimstepf/Dev/web-app/log/passenger.5000.log
11:05:03 web.1    | Environment: development
11:05:03 web.1    | Accessible via: http://0.0.0.0:5000/
11:05:03 web.1    | 
11:05:03 web.1    | You can stop Phusion Passenger Standalone by pressing Ctrl-C.
11:05:03 web.1    | Problems? Check https://www.phusionpassenger.com/documentation/Users%20guide%20Standalone.html#troubleshooting
11:05:03 web.1    | ===============================================================================
11:05:06 web.1    | App 1348 stdout: 
11:05:11 web.1    | App 1348 stdout: 
11:05:11 web.1    | [ 2015-05-01 11:05:11.4894 1329/0x109438000 App/Implementation.cpp:287 ]: Could not spawn process for application /Users/michaelimstepf/Dev/web-app: An error occured while starting up the preloader.
11:05:11 web.1    |   Error ID: fb5bc44c
11:05:11 web.1    |   Error details saved to: /var/folders/dy/rqh4m_713fx624wgtlvcwy840000gn/T//passenger-error-bgFXQJ.html
11:05:11 web.1    |   Message from application: Address already in use - bind(2) for "localhost" port 5051 (Errno::EADDRINUSE)
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/byebug-4.0.5/lib/byebug/remote.rb:52:in `initialize'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/byebug-4.0.5/lib/byebug/remote.rb:52:in `new'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/byebug-4.0.5/lib/byebug/remote.rb:52:in `start_control'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/byebug-4.0.5/lib/byebug/remote.rb:31:in `start_server'
11:05:11 web.1    |   /Users/michaelimstepf/Dev/web-app/config/environments/development.rb:6:in `block in <top (required)>'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.1/lib/rails/railtie.rb:210:in `instance_eval'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.1/lib/rails/railtie.rb:210:in `configure'
11:05:11 web.1    |   /Users/michaelimstepf/Dev/web-app/config/environments/development.rb:4:in `<top (required)>'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activesupport-4.2.1/lib/active_support/dependencies.rb:274:in `require'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activesupport-4.2.1/lib/active_support/dependencies.rb:274:in `block in require'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activesupport-4.2.1/lib/active_support/dependencies.rb:240:in `load_dependency'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activesupport-4.2.1/lib/active_support/dependencies.rb:274:in `require'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.1/lib/rails/engine.rb:598:in `block (2 levels) in <class:Engine>'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.1/lib/rails/engine.rb:597:in `each'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.1/lib/rails/engine.rb:597:in `block in <class:Engine>'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.1/lib/rails/initializable.rb:30:in `instance_exec'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.1/lib/rails/initializable.rb:30:in `run'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.1/lib/rails/initializable.rb:55:in `block in run_initializers'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/2.2.0/tsort.rb:226:in `block in tsort_each'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/2.2.0/tsort.rb:348:in `block (2 levels) in each_strongly_connected_component'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/2.2.0/tsort.rb:420:in `block (2 levels) in each_strongly_connected_component_from'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/2.2.0/tsort.rb:429:in `each_strongly_connected_component_from'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/2.2.0/tsort.rb:419:in `block in each_strongly_connected_component_from'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.1/lib/rails/initializable.rb:44:in `each'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.1/lib/rails/initializable.rb:44:in `tsort_each_child'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/2.2.0/tsort.rb:413:in `call'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/2.2.0/tsort.rb:413:in `each_strongly_connected_component_from'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/2.2.0/tsort.rb:347:in `block in each_strongly_connected_component'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/2.2.0/tsort.rb:345:in `each'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/2.2.0/tsort.rb:345:in `call'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/2.2.0/tsort.rb:345:in `each_strongly_connected_component'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/2.2.0/tsort.rb:224:in `tsort_each'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/2.2.0/tsort.rb:203:in `tsort_each'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.1/lib/rails/initializable.rb:54:in `run_initializers'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.1/lib/rails/application.rb:352:in `initialize!'
11:05:11 web.1    |   /Users/michaelimstepf/Dev/web-app/config/environment.rb:5:in `<top (required)>'
11:05:11 web.1    |   config.ru:3:in `require'
11:05:11 web.1    |   config.ru:3:in `block in <main>'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rack-1.6.0/lib/rack/builder.rb:55:in `instance_eval'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rack-1.6.0/lib/rack/builder.rb:55:in `initialize'
11:05:11 web.1    |   config.ru:1:in `new'
11:05:11 web.1    |   config.ru:1:in `<main>'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/passenger-5.0.6/helper-scripts/rack-preloader.rb:111:in `eval'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/passenger-5.0.6/helper-scripts/rack-preloader.rb:111:in `preload_app'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/passenger-5.0.6/helper-scripts/rack-preloader.rb:157:in `<module:App>'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/passenger-5.0.6/helper-scripts/rack-preloader.rb:29:in `<module:PhusionPassenger>'
11:05:11 web.1    |   /Users/michaelimstepf/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/passenger-5.0.6/helper-scripts/rack-preloader.rb:28:in `<main>'
11:05:11 web.1    | 
11:05:11 web.1    | 
11:05:11 web.1    | [ 2015-05-01 11:05:11.4980 1329/0x109122000 age/Hel/Req/CheckoutSession.cpp:252 ]: [Client 1-1] Cannot checkout session because a spawning error occurred. The identifier of the error is fb5bc44c. Please see earlier logs for details about the error.
```

### LINKED ISSUE #239: Cannot interrupt program execution from remote control interface
After sending PR https://github.com/deivid-rodriguez/byebug/pull/238 to fix remote debugging, I still cannot interrupt program execution from the remote control interface.

I'm starting a simple Ruby script that loops forever with a 5 second delay per loop and a calls 'puts' with the loop count. I run it with remote byebug and I connect to the debugging remote and control ports. Everything looks fine, but when I issue an interrupt command in the control port while the program is running, it replies that Byebug has not been started yet.

I'm not sure what this means. If it cannot be fixed now, any hint on how to continue diagnosing/fixing this will be appreciated. 

Thanks in advance.

### LINKED ISSUE #274: Remote Debugger Hanging Up Crashes Byebug
**Problem:** If the socket for a remote connection is closed after connecting to the remote server, further writes/reads will crash the server.

**Version:** 9.05

**Steps to Reproduce**:
 1) Run this example program

```
require 'byebug'
require 'byebug/core'

Byebug.wait_connection = true
Byebug.start_server 'localhost', ENV.fetch("BYEBUG_SERVER_PORT", 1048).to_i

puts 'a'
byebug
puts 'b'
sleep 5
puts 'c'
byebug
puts 'd'
```

2) Connect to this using the remote debugger
3) When it encounters the first `byebug` statement, `continue`
4) While the program sleeps, terminate the byebug CLI (CTRL + C)
5) After the program wakes up, when it hits the second `byebug` statement, it will crash with an `Errno::EPIPE` exception

**Expectation**:  The server should either detect when the socket is closed or if an EPIPE exception has happened, and stop communicating over the socket (treat it as if there was no connection).

While the above example is very arbitrary, a more real world example of where this happens: debugging remote rails applications. If you have a byebug statement in a controller, connect to it remotely, debug the area, then the debugger closes for some reason you will have a crash.

Fixing this will also allow users to "detach" the debugger by interrupting it between byebug calls (useful if you have a constantly running dev server).

### LINKED ISSUE #302: Slow debugger in remote mode
## Problem description

It seems that Byebug.start_server at the top of my (sinatra) application causes near 30x performance regressions in some cases.

A certain query which fires off a bunch of DataMapper queries and returns a ton of JSON executes approximately 30x slower when a byebug server is started (with no clients connected)

## Expected behavior

I would expect no performance penalty, particularly if no client is connected.

## Actual behavior

The given endpoint benchmarks (with wrk) 30x slower.

## Steps to reproduce the problem

Unfortunately, the code isn't open source. I can try to create a small reproduction scenario, first, though, I was wondering if this is known/expected? Perhaps my assumptions about performance are incorrect?

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
Fixes #82.
Fixes #141.
Fixes #239.
Fixes #274.
Fixes #302.

### CODE REVIEW

**Comment by deivid-rodriguez** [comment]
Summoning @MSP-Greg in case you're interested. I've been working on remote debugging support. Not ready yet but looking good. However, I'm getting some failures on Windows and I don't have a Windows machine to debug. In case you would like to help out, I suspect it's something around setting binary mode in streams, but not sure. Regards!

**Comment by MSP-Greg** [comment]
@deivid-rodriguez

I'll have a look at it, might not be able to seriously do so until Saturday...  Thanks, Greg

**Comment by deivid-rodriguez** [comment]
That's great, thanks so much!

**Comment by MSP-Greg** [comment]
@deivid-rodriguez

Solution found.  See  [commit in my fork](https://github.com/MSP-Greg/byebug/commit/9ef60313c61b58c97c72c6f18b7b528c6021e42d), [Appveyor results](https://ci.appveyor.com/project/MSP-Greg/byebug)

Probably not what you (or I) expected.  Hope this helps, Greg

**Comment by MSP-Greg** [comment]
Added another patch to swap the fork for Thread.  See [commit in my fork](https://github.com/MSP-Greg/byebug/commit/ed8205b8261cf9ffe175f7d785724aa4cd68164f).  [Appveyor results](https://ci.appveyor.com/project/MSP-Greg/byebug/build/2).

Appveyor runs with dual cores, Travis has four.  The `sleep(windows? ? 2 : 1)` line may need windows increased to 2.5 or 3, not sure.  The first build I ran stopped with a very strange log, second build passed...

**Comment by deivid-rodriguez** [comment]
@MSP-Greg Thanks so much! Your fix makes a lot of sense since only `byebug`'s executable [sets up the LOAD_PATH](https://github.com/deivid-rodriguez/byebug/blob/1b55351e045ed67cd3e87493a5996ccb77a77f4c/exe/byebug#L4-L5) but it can't be used directly on Windows!

Also thanks for letting me know about the ruby-core issue, I commented in there! And also thanks for fixing those tests to not use `fork`.

I'll either give you access to this branch to add your changes, or cherry-pick them myself!

**Comment by MSP-Greg** [comment]
@deivid-rodriguez

> Also thanks for letting me know about the ruby-core issue, I commented in there!

Thank you.  I keep an eye on it (and the ruby-core Travis & Appveyor), and I knew you could address the issue(s) well.

I'm also sensitive to issues that involve trunk, If someone is concerned about a given repo/gem functioning with trunk, they're welcome to fork and run their own CI, since many 'mature' repos/gems may not have frequent commits...

> I'll either give you access to this branch to add your changes, or cherry-pick them myself!

Whatever is easiest.

Finally, I reverted 'Binmode?' with two commits, both passed.

Thanks, Greg

**Comment by MSP-Greg** [comment]
@deivid-rodriguez

Would you prefer that I delete or revert the `Binmode?` commit (since it is the last one)?

**Comment by deivid-rodriguez** [comment]
Delete :+1:

**Comment by MSP-Greg** [comment]
Sometimes Appveyor...  Had this issue last night, where I changed `sleep 1` to `sleep(windows? ? 2 : 1)`.

Just upped it to 3, no change.  I'm going to check/fix it on my fork.  I must be missing something....

FYI - from all my testing on Appveyor (100's of trunk builds, rubygems builds, etc), I've seen lots of intermittent issues...

**Comment by MSP-Greg** [comment]
@deivid-rodriguez

I found what the issue was with my fork passing the one test.  I dl'd a diff of the fork, and (oddly) one line was different.  I even checked my browser history, and timestamps matched, so it was what GitHub had.  The line [`def write_program(code)`](https://github.com/deivid-rodriguez/byebug/blob/fix/remote_debugging/test/remote_test.rb#L138) was changed to `def write_program(program)`.

Anyway, fixed that, I think I've got this working, juts a few minutes.  The change is to [`RemoteInterface#readline`](https://github.com/deivid-rodriguez/byebug/blob/7bddb3e688d1a203f962c2b3570974a50abc2771/lib/byebug/interfaces/remote_interface.rb#L41-L48)

to this:

```ruby
def readline(prompt)
  puts(prompt)

  result = (input.gets || "continue")
  result.chomp
rescue Errno::EPIPE, Errno::ECONNABORTED
  "continue"
end
```

`Errno::EPIPE` is raised by trunk, `Errno::ECONNABORTED` is raised by 2.4 and lower.

Thoughts?  trying on my fork shortly...  Thanks, Greg

**Comment by deivid-rodriguez** [comment]
> I found what the issue was with my fork passing the one test. I dl'd a diff of the fork, and (oddly) one line was different. I even checked my browser history, and timestamps matched, so it was what GitHub had. The line def write_program(code) was changed to def write_program(program).

Yeah, sorry! I realized that was incorrect and fixed it here, leaving your fork up to date indeed. Feel free to just work on this same branch instead of your fork, I'll be carefull with force-pushing :)

Your fix looks good, but I think the `rescue` should be moved up to the `confirm` and `read_command` methods (which internally call `readline`), similarly to `puts` and `print`. Otherwise I think the exception could be raised outside of the `rescue` block you're adding. What do you think?

**Comment by MSP-Greg** [comment]
> Yeah, sorry!

That was before I got changed to 'collaborator', so I grabbed the changes and was working just in my fork in a different new branch.  IOW, no problem.

> I'll be carefull with force-pushing :)

I do that alot also.  I'll try to restrict it to things I do **before** I push here.

> I think the rescue should be moved up to the confirm and read_command methods (which internally call readline), similarly to puts and print. Otherwise I think the exception could be raised outside of the rescue block you're adding. What do you think?

Well, it (as listed) just [passed](https://ci.appveyor.com/project/MSP-Greg/byebug/build/7)  on my fork with Appveyor.  The commit is [here](https://github.com/MSP-Greg/byebug/commit/ce1acad6a4122517fadf63cba8a6eb7614ae27a5).

I haven't reviewed the remote code enough to know if a particular error should trigger a 'is the client  dead?' check,..

Lastly, can we please remove the 'stop on failure' from Appveyor?  I normally do all of my work with trunk.  I was passing locally, pushed to Appveyor, and the build stopped on 2.3 (which lead to my adding `Errno::ECONNABORTED`).

**Comment by deivid-rodriguez** [comment]
> Well, it (as listed) just passed on my fork with Appveyor. The commit is here.

Yeah, we can't know where the exception will be raised, and most of the times it will happen under the `rescue` block, but my alternative should be safer I think.

> Lastly, can we please remove the 'stop on failure' from Appveyor? I normally do all of my work with trunk. I was passing locally, pushed to Appveyor, and the build stopped on 2.3 (which lead to my adding Errno::ECONNABORTED).

Ok, since you are the only one here working on Windows, let's do whatever works best for you :)

**Comment by MSP-Greg** [comment]
> Yeah, we can't know where the exception will be raised, and most of the times it will happen under the rescue block, but my alternative should be safer I think.

I'll defer to you.  I'm just starting at the tests and moving inward.  You know the whole thing.

Changed. Test [passed](https://ci.appveyor.com/project/MSP-Greg/byebug/build/8).  Commit is [here](https://github.com/MSP-Greg/byebug/commit/32fe8eed5e82b2a6ada89919102527782d31908f).  I changed the `readline` code to remove the `result` variable.  Old habit, I prefer to never use local variables if not needed, because 6 months from now, I'd feel a need to double check if `result` was an attribute I'd forgotten about.  I think that made sense....

Let me know if it's good, and I'll push it.  then I'll move on to #416...

**Comment by deivid-rodriguez** [comment]
> Changed. Test passed. Commit is here.

Yep, that's what I meant. However, the `confirm` part should probably default to `false`, not `"continue"` since `super` returns a boolean. That said, that part is untested at the moment, so you can just leave it alone for now. So I'd say no change at all there, or rescue and return `false`, up to you!

> I changed the readline code to remove the result variable. Old habit, I prefer to never use local variables if not needed, because 6 months from now, I'd feel a need to double check if result was an attribute I'd forgotten about. I think that made sense..

Mmmm, not sure. I think local variables have preference over attributes, if you were to set an attribute, you'd have to do `self.result =`, I think? The new code is more concise but unnecessarily `chomps` "continue". So I'm "meh" on the change, but let's do it if you prefer it that way.

**Comment by MSP-Greg** [comment]
> rescue and return false, up to you!

Sorry I missed that.  Changed.

>  unnecessarily chomps "continue".

Can't chomp nil though...

Pushing shortly.  Thanks.

**Comment by deivid-rodriguez** [comment]
> Sorry I missed that. Changed.

:+1: 

> > unnecessarily chomps "continue".
>
> Can't chomp nil though...

I mean that, correctly but unnecessarily, calls `"continue".chomp` (which returns "continue" again) when `input.gets` is `nil`. I do like its simplicity though, let's do it! :+1:

**Comment by MSP-Greg** [comment]
Cool. Sorry, I meant that `input.gets.chomp || "continue"` could error.

Thanks for cancelling Appveyor...

**Comment by deivid-rodriguez** [comment]
> Cool. Sorry, I meant that input.gets.chomp || "continue" could error.

Yup, that could fail. It was you way or mine, yours works :)

> Thanks for cancelling Appveyor...

No problem.

I'll give this a final look tomorrow, hopefully I can add coverage for a couple more cases!

**Comment by deivid-rodriguez** [comment]
@MSP-Greg I just rebased this PR!

I had also added a some more tests and refactorings to the remote debugging specs but you know what? It made it hang again. :rage:

That gave me a clue though to further isolate the bug / problem with `Open3` + `rb-readline`. Maybe it happens when you use threads inside `popen2e`'s block? One difference between the current state of this PR and the new tests I had added or what we are doing in #416 is that we currently don't do that here... Note that [`capture2e` implementation](https://github.com/ruby/ruby/blob/3d9453994c965abff8851d008a01b40153ffee77/lib/open3.rb#L348-L368), which we are using in #416 uses `Thread` inside `popen2e`'s block...

**Comment by MSP-Greg** [comment]
@deivid-rodriguez

Just looked, the last AV run passed.  Things good now?

Re `Open3`, I've got the readline prompt appearing, but after that, it goes somewhere unknown.  I've tried with and without threads.   Of course, when I switch to `readline.so`, the code runs fine...

**Comment by deivid-rodriguez** [comment]
Rebased and added new tests and refactors and the Windows build passed! :tada:

File: lib/byebug/interfaces/remote_interface.rb [hunk 1]
```diff
@@ -16,25 +16,35 @@ def initialize(socket)
...
 
     def confirm(prompt)
       super("CONFIRM #{prompt}")
+    rescue Errno::EPIPE, Errno::ECONNABORTED
+      false
+    end
+
+    def print(message)
+      super(message)
+    rescue Errno::EPIPE
```
Comment (deivid-rodriguez) [inline lib/byebug/interfaces/remote_interface.rb]: @MSP-Greg Should we rescue `Errno::ECONNABORTED` here and in `puts` as well? You were seeing that exception on Windows, but only under some versions if I recall correctly?

Comment (MSP-Greg) [reply lib/byebug/interfaces/remote_interface.rb]: I don't recall, I'll have to check.

Comment (deivid-rodriguez) [reply lib/byebug/interfaces/remote_interface.rb]: You mentioned it [here](https://github.com/deivid-rodriguez/byebug/pull/406#issuecomment-355767858). I think I'll add `Errno::ECONNABORTED` in the other methods as well then.

Re `exe/byebug`, feel free to suggest any changes you want in a separate PR!

Comment (MSP-Greg) [reply lib/byebug/interfaces/remote_interface.rb]: > You mentioned it here.

Sorry, rather distracted with `non code` things yesterday evening.

Travis has had large Linux & MacOS backlogs recently (separate occasions).  Hope they get it sorted out...

**Comment by deivid-rodriguez** [comment]
Let's do this!

**Pull request merged**